### PR TITLE
[WIP] setup.py changes to include ansible-test

### PR DESCRIPTION
**Pull request by nitzmahone**
##### SUMMARY
* stop using `package_data`
* update package-data sanity test to ensure everything's getting installed that's in MANIFEST
* add ansible-test entrypoint
* name package groups
* add test/lib/ansible_test

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-test

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -275,6 +275,14 @@ def get_dynamic_setup_params():
             'config/base.yml',
             'config/module_defaults.yml',
         ],
+        'ansible_test': [
+            '*/*.*',
+            '*/*/*.*',
+            '*/*/*/*.*',
+            '*/*/*/*/*.*',
+            '*/*/*/*/*/*.*',
+            '*/*/*/*/*/*/*.*'
```
Comment (mattclay) [inline setup.py]: Files without extensions need to be included as well. For example, all the files in this directory:

https://github.com/ansible/ansible/tree/devel/test/lib/ansible_test/_data/sanity/pylint/config

File: setup.py [hunk 2]
```diff
@@ -275,6 +275,14 @@ def get_dynamic_setup_params():
...
+        ]
```
Comment (sivel) [inline setup.py]: We should update `test/sanity/code-smell/package-data.py` to ensure things get installed.  That `package_data` looks like we could end up with missing files easily if nesting depth changes.

File: setup.py (hunk 1 above)
Comment (mattclay) [reply setup.py]: We'll also want an `exclude_package_data` entry to exclude all of `test/lib/ansible_test/tests/`.

**Comment by abadger** [comment]
What is the goal here?  By goal, I mean, what do you want on the filesystem, in what locations, for the installed machine?  There might be an easier way to achieve that (alternately, it might be much harder than this).

Also, some of these things might have undesirable side-effects.  I think that @sivel fixed a bug at one point by messing with package_dir.  Someone will have to look through git history to make sure we aren't regressing that fix.

File: setup.py (hunk 1 above)
Comment (abadger) [reply setup.py]: I think you probably want a function which does this rather than trying to hardcode wildcarded strings in here.

**Comment by abadger** [comment]
It wasn't sivel:   https://github.com/ansible/ansible/issues/10437  That issue has a reproducer as well so we can test.

**Comment by mattclay** [comment]
@abadger The goal is for `test/lib/ansible_test/` to be installed as the package `ansible_test` alongside the `lib/ansible/` content available as `ansible`.

File: setup.py (hunk 2 above)
Comment (nitzmahone) [reply setup.py]: I don't actually think we have a good reason to be using `package_data` at all anymore- we should be using `include_package_data=True` and setting everything up in `MANIFEST.in`. I'm updating the PR to do that (for ansible as well) and updating the sanity test to validate the installed bits for both against MANIFEST.

File: setup.py (hunk 1 above)
Comment (nitzmahone) [reply setup.py]: handled this a different way: ditched package_data altogether and updated sanity test accordingly

**Comment by nitzmahone** [comment]
@abadger I've done some side-by-side tests against devel- as of the latest commit, which ditches package_data altogether in favor of include_package_data, it's including the same stuff in the sdist and installed bits that it was before (plus of course all the ansible-test stuff). Ditching package_data also gets rid of a growing maintenance hassle- basically if it's under the ansible or ansible_test packages, it gets shipped. We can use `exclude_package_data` if we *really* need to, but at the moment, there shouldn't be a need.

**Comment by nitzmahone** [comment]
@abadger Hadn't refreshed to see your comment and PR- this *does* seem to regress #10437, but I'm wondering if there's a simpler solution that lets us leave all the logic in the manifest and setuptools as it does now...

**Comment by abadger** [comment]
@nitzmahone <nod>  You may be able to avoid the regression by simply not changing the key for empty string to 'ansible' in the package_dir entry.

According to mattclay, we will need to exclude some data files (everything in ansible_test/tests).  I'm not sure whether exclude_package_data will work for that... it might if you can specify a directory prefix like
```
exclude_package_data={'ansible_test': 'tests/*'}
```

Since, most other things in setup.py aren't recursive, you may need to write a function to walk the tree for this.  That would likely be a simper function than the function which enumerates what to include (since that one needs to know about both what to include and what to exclude.)

**Comment by abadger** [comment]
Looking at the way the test is changed here, I don't think we should rely on MANIFEST in order to determine the file list that we want.  MANIFEST is generated as part of the build process and, when using include_package_data=True is the basis for determining what setup.py is going to install.  So relying upon MANIFEST in the test won't capture any problems where we forget to update MANIFEST.in or what's in setup.py.

Previously our issue was that we were specifying the files to include manually and forgetting to install those.  If you're able to adapt exclude_package_data, then our problem will be that we are specifying exclude_package_data manually, so we might have to change the test to be more worried about failures to exclude.

Separately, I don't think we should add --no-compile to the setup.py args.  People won't be using --no-compile to install so it feels like we should do that filtering inside of our code rather than modifying the steps that get the files the users will see.  I'm on the fence about --prefix.... It is similar to --root as it is just modifying the directory paths and we were already getting info from the install itself about what prefix was being used but it also isn't commonly used by people installing ansible.  (--root is used by distros when packaging so that seems to fall on the okay to use here line).

**Comment by nitzmahone** [comment]
As it is pretty much every time I have to touch Python packaging, I'm > 12h of futzing with something I was hoping would take 15-20m, so I don't really care what we do at this point. `ansible-test` needs to be installed as a top-level package, and hopefully we're not making things significantly more complex than they already are. If @mattclay is happy with what you have in #60718, just kill this and merge that. I needed to move on ~11h of hassle ago.

**Pull request closed without merging**